opentracing-api-extensions-tracer/src/main/java/io/opentracing/contrib/api/tracer/APIExtensionsSpan.java

```diff
@@ -0,0 +1,282 @@
+
+public class APIExtensionsSpan implements Span, SpanData  {
+
+    private final Span wrappedSpan;
+
+    private String operationName;
+    private final long startTimestampMicro;
+    private long finishTimestampMicro;
+    private final long startTimeNano;
+    private long finishTimeNano;
+    private final Map<String,Object> tags;
+
+    private final List<SpanObserver> observers = new CopyOnWriteArrayList<SpanObserver>();
+
+    private final UUID correlationId = UUID.randomUUID();
+
+    /**
+     * This is the constructor for the extensions API span wrapper.
+     *
+     * @param span The span being wrapped
+     * @param operationName The operation name
+     * @param startTimestampMicro The start timestamp (microseconds)
+     * @param startTimeNano The start nano time, or 0 if the start timestamp was explicitly provided by the app
+     * @param tags The initial tags
+     */
+    public APIExtensionsSpan(Span span, String operationName,
+            long startTimestampMicro, long startTimeNano, Map<String,Object> tags) {
+        this.wrappedSpan = span;
+        this.operationName = operationName;
+        this.startTimestampMicro = startTimestampMicro;
+        this.startTimeNano = startTimeNano;
+        this.tags = tags;
+    }
+
+    /**
+     * This method adds a new {@link SpanObserver}.
+     *
+     * @param observer The observer
+     */
+    public void addSpanObserver(SpanObserver observer) {
+        if (observer != null) {
+            observers.add(observer);
+        }
+    }
+
+    /**
+     * This method removes a {@link SpanObserver}.
+     *
+     * @param observer The observer
+     */
+    public void removeSpanObserver(SpanObserver observer) {
+        if (observer != null) {
+            observers.remove(observer);
+        }
+    }
+
+    @Override
+    public SpanContext context() {
+        return wrappedSpan.context();
+    }
+
+    @Override
+    public Object getCorrelationId() {
+        return correlationId;
+    }
+
+    @Override
+    public long getStartTime() {
+        return startTimestampMicro;
+    }
+
+    @Override
+    public long getFinishTime() {
+        return finishTimestampMicro;
+    }
+
+    @Override
+    public Span setOperationName(String operationName) {
+        wrappedSpan.setOperationName(operationName);
+        this.operationName = operationName;
+        if (!observers.isEmpty()) {
+            for (SpanObserver observer : observers) {
+                observer.onSetOperationName(this, operationName);
+            }
+        }
+        return this;
+    }
+
+    @Override
+    public String getOperationName() {
+        return operationName;
+    }
+
+    @Override
+    public String getBaggageItem(String name) {
+        return wrappedSpan.getBaggageItem(name);
+    }
+
+    @Override
+    public Span setBaggageItem(String name, String value) {
+        wrappedSpan.setBaggageItem(name, value);
+        if (!observers.isEmpty()) {
```

Comment:
(there are other places like this, so, if you do change this, the other places might need changing as well)